### ISSUE TITLE
CheckRules before Insert/Update?

### ISSUE THREAD

**Issue by ajj3085** [MarimerLLC/cslaforum#229]
I'm working on getting my first Xamarin mobile app going using Csla, and will use a remote data portal so that the objects run on both the mobile app and the asp.net server.  For some reason, "don't trust the client" popped into my head, and i know with mobile apps its pretty easy to get to their source (APKs seem trivial to convert to Java code). 

In such a scenario, would calling BusinessRules.CheckRules at the top of the DataPortal_Insert/Update be advisable, or is this being overly paranoid?  What is commonly done?

**Comment by rockfordlhotka** [comment]
Yes, if you don't trust the client it is a good idea to ensure all rules run (again) on the server where the code is for sure under your control.

As with all security concerns you should create a threat matrix to decide the likelihood, cost-of-breach, and cost-of-mitigation for each potential threat so you can decide whether the cost of mitigation is worth it compared to the cost/likelihood of a breach.

In terms of making the call at the top of each DP_Update method, that's fine. If you are worried about a developer forgetting to make such a call you might consider making the call in the server-side data portal activation chain so it is a guaranteed global step - perhaps in the authorization step, where you might not allow processing to continue if the root object isn't valid.

**Comment by ajj3085** [comment]
Thanks, some great things to consider!

**Comment by fujiiface** [comment]
@ajj7060 Did you end up implementing this in the authorization step or at the top of your DataPortal_Update/Insert?

**Comment by jonnybee** [comment]
You should also consider to configure any async business rules (ex Lookup rules) so that they can not run on the "logical serverside" and/or in BusinessRules.CheckRules.

**Comment by ajj7060** [comment]
@fujiiface For the use case I ended up not doing either as the risk seems minimal (the queries filter based on current user permissions, which are kept server side and not modifiable from the client).  If I did need to do this I'd probably call CheckRules at the top of the DP_Insert/Update, as I'd like to make the decision on a case-by-case basis.
